### PR TITLE
Route creation reconciler loop.

### PR DESCRIPTION
**Pull request by cjcullen**
This teases apart node registration from cloudprovider routing configuration. It also makes route configuration robust against controller crashes and transient cloudprovider failures.

RIght now, I just have this naively syncing every 10 seconds. It would be nicer to react to changes in the NodeList, and then just run a background sync every few minutes. (@lavalamp, I'm told you're the expert on that :) )

I'd like to wait for #7984 to get in and then rebase this before merging.

### CODE REVIEW

File: pkg/cloudprovider/gce/gce.go [hunk 1]
```diff
@@ -555,31 +560,45 @@ func getMetadataValue(metadata *compute.Metadata, key string) (string, bool) {
...
 		Network:         fmt.Sprintf("global/networks/%s", gce.networkName),
 		Priority:        1000,
+		Description:     route.Description,
 	}).Do()
 	if err != nil {
 		return err
 	}
-	if err := gce.waitForGlobalOp(insertOp); err != nil {
-		if gapiErr, ok := err.(*googleapi.Error); ok && gapiErr.Code == http.StatusConflict {
-			// TODO (cjcullen): Make this actually check the route is correct.
```
Comment (cjcullen) [inline pkg/cloudprovider/gce/gce.go]: The resolution of this TODO is that the reconciler will now remove an incorrect route and create a correct one. Any transient 409s or 404s from reconciling will be logged, but will not prevent progress.

File: pkg/cloudprovider/routecontroller/routecontroller.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+	return rc.clusterName
+}
+
+func (rc *RouteController) isResponsibleForRoute(route *cloudprovider.Route) bool {
+	_, cidr, err := net.ParseCIDR(route.DestinationCIDR)
+	if err != nil {
+		glog.Errorf("Ignoring route %s, unparsable CIDR: %v", route.Name, err)
+		return false
+	}
+	// Not responsible if this route's CIDR is not within our clusterCIDR
```
Comment (davidopp) [inline pkg/cloudprovider/routecontroller/routecontroller.go]: is -> if

**Comment by davidopp** [comment]
(Other than that, LGTM.)

File: pkg/cloudprovider/routecontroller/routecontroller.go (hunk 2 above)
Comment (cjcullen) [inline pkg/cloudprovider/routecontroller/routecontroller.go]: I think the comment is grammatically correct. Sounds like the wording might be a bit confusing. I'm open to suggestions.

File: pkg/cloudprovider/routecontroller/routecontroller_test.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+		routes, ok := cloud.Routes()
+		if !ok {
+			t.Error("Error in test: fake_cloud doesn't support Routes()")
+		}
+		_, cidr, _ := net.ParseCIDR("10.120.0.0/16")
+		rc := New(routes, nil, cluster, cidr)
+		if err := rc.reconcile(testCase.nodes, testCase.initialRoutes); err != nil {
+			t.Errorf("%d. Error from rc.reconcile(): %v", i, err)
+		}
+		time.Sleep(10 * time.Millisecond)
```
Comment (cjcullen) [inline pkg/cloudprovider/routecontroller/routecontroller_test.go]: I feel gross about this. The reconciler kicks off goroutines to add/delete routes in the cloudprovider, so immediately checking the state of the fake_cloud in the main thread will show that nothing has been updated. A sleep of pretty much any length is enough to allow the background threads to update the in-memory representation, but it feels so wrong.

Any suggestions on a better way?

**Comment by dchen1107** [comment]
@cjcullen Have you run  e2e test with your change yet?

**Comment by cjcullen** [comment]
Yes. All green (1 time). This was already rebased on top of #6949 though :/

File: pkg/cloudprovider/routecontroller/routecontroller_test.go (hunk 3 above)
Comment (quinton-hoole) [inline pkg/cloudprovider/routecontroller/routecontroller_test.go]: Notify via a channnel that one is complete.

**Comment by cjcullen** [comment]
Rebased against #6949. Now running e2e's (over and over again).

**Comment by cjcullen** [comment]
2 for 2 on e2es so far after the rebase (tearing down and rebuilding each time)

File: pkg/cloudprovider/routecontroller/routecontroller.go (hunk 2 above)
Comment (davidopp) [inline pkg/cloudprovider/routecontroller/routecontroller.go]: Sorry, it actually makes total sense, I just mis-read it.

**Comment by cjcullen** [comment]
This already got an LGTM before I rebased. Any opposition to me adding the status/LGTM tag?

File: pkg/cloudprovider/routecontroller/routecontroller.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+		}
+	}, syncPeriod)
+}
+
+func (rc *RouteController) reconcileNodeRoutes() error {
+	routeList, err := rc.routes.ListRoutes(rc.truncatedClusterName() + "-.*")
+	if err != nil {
+		return fmt.Errorf("error listing routes: %v", err)
+	}
+	nodeList, err := rc.kubeClient.Nodes().List(labels.Everything(), fields.Everything())
```
Comment (lavalamp) [inline pkg/cloudprovider/routecontroller/routecontroller.go]: // TODO: use pkg/controller/framework.NewInformer to watch this and reduce the number of lists needed.

File: pkg/cloudprovider/nodecontroller/nodecontroller.go [hunk 5]
```diff
@@ -567,6 +340,9 @@ func (nc *NodeController) monitorNodeStatus() error {
 	if err != nil {
 		return err
 	}
+	if nc.allocateNodeCIDRs {
+		nc.reconcileNodeCIDRs(nodes)
```
Comment (lavalamp) [inline pkg/cloudprovider/nodecontroller/nodecontroller.go]: // TODO: use stuff from the controller framework to watch, to reduce the number of lists executed.

File: pkg/cloudprovider/cloud.go [hunk 6]
```diff
@@ -80,12 +83,27 @@ type Instances interface {
...
-	// Configure the specified instance using the spec
-	Configure(name string, spec *api.NodeSpec) error
-	// Delete all the configuration related to the instance, including other cloud resources
-	Release(name string) error
 }
 
+// Route is a representation of an advanced routing rule.
+type Route struct {
+	Name            string
+	TargetInstance  string
```
Comment (lavalamp) [inline pkg/cloudprovider/cloud.go]: Comment on this-- is it the hostname, ip address, or name-as-given-by-cloud-provider?

File: cluster/gce/util.sh [hunk 7]
```diff
@@ -772,7 +772,7 @@ function kube-down {
   # Delete routes.
   local -a routes
   routes=( $(gcloud compute routes list --project "${PROJECT}" \
-              --regexp "${NODE_INSTANCE_PREFIX}-.+" | awk 'NR >= 2 { print $1 }') )
+    --regexp "${INSTANCE_PREFIX}-.{8}-.{4}-.{4}-.{4}-.{12}" | awk 'NR >= 2 { print $1 }') )
```
Comment (lavalamp) [inline cluster/gce/util.sh]: What does this line do? (please add a comment)

**Comment by lavalamp** [comment]
LGTM-- a few requests for comments

File: pkg/cloudprovider/routecontroller/routecontroller.go (hunk 4 above)
Comment (cjcullen) [inline pkg/cloudprovider/routecontroller/routecontroller.go]: Done.

File: pkg/cloudprovider/nodecontroller/nodecontroller.go (hunk 5 above)
Comment (cjcullen) [inline pkg/cloudprovider/nodecontroller/nodecontroller.go]: Done.

File: pkg/cloudprovider/cloud.go (hunk 6 above)
Comment (cjcullen) [inline pkg/cloudprovider/cloud.go]: Done.

File: cluster/gce/util.sh (hunk 7 above)
Comment (cjcullen) [inline cluster/gce/util.sh]: Done.

**Comment by lavalamp** [comment]
Thank you very much, unfortunately I noticed one more thing. Sorry!

File: pkg/cloudprovider/routecontroller/routecontroller_test.go [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+		if err := rc.reconcile(testCase.nodes, testCase.initialRoutes); err != nil {
+			t.Errorf("%d. Error from rc.reconcile(): %v", i, err)
+		}
+		var finalRoutes []*cloudprovider.Route
+		var err error
+		timeoutChan := time.After(50 * time.Millisecond)
+		tick := time.Tick(10 * time.Millisecond)
+	poll:
+		for {
+			select {
```
Comment (cjcullen) [inline pkg/cloudprovider/routecontroller/routecontroller_test.go]: @lavalamp is this more like what you were thinking?

Comment (lavalamp) [inline pkg/cloudprovider/routecontroller/routecontroller_test.go]: Yes, thank you, much better!

File: pkg/cloudprovider/routecontroller/routecontroller_test.go [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+		}
+		_, cidr, _ := net.ParseCIDR("10.120.0.0/16")
+		rc := New(routes, nil, cluster, cidr)
+		if err := rc.reconcile(testCase.nodes, testCase.initialRoutes); err != nil {
+			t.Errorf("%d. Error from rc.reconcile(): %v", i, err)
+		}
+		var finalRoutes []*cloudprovider.Route
+		var err error
+		timeoutChan := time.After(50 * time.Millisecond)
+		tick := time.Tick(10 * time.Millisecond)
```
Comment (lavalamp) [inline pkg/cloudprovider/routecontroller/routecontroller_test.go]: I think you need to stop or delete this when you're done with it, though?

Comment (cjcullen) [inline pkg/cloudprovider/routecontroller/routecontroller_test.go]: Ahh. You're absolutely right. How's that?

**Comment by lavalamp** [comment]
LGTM

**Comment by dchen1107** [comment]
LGTM. but still @cjcullen have you run e2e with this one yet?

cc/ @justinsb on aws, and @derekwaynecarr on vagrant. Just a FYI.

**Comment by cjcullen** [comment]
e2e's were solid green 2 days ago. I'll rebase and run another round (or 3) to get a little more confidence that this will play nice with what we have at head.

**Comment by dchen1107** [comment]
@cjcullen Thanks for your patience.

**Pull request merged**